mobile/library/java/org/chromium/net/impl/CronvoyUrlRequest.java

```diff
@@ -864,10 +863,20 @@ public void onData(ByteBuffer data, boolean endStream, EnvoyStreamIntel streamIn
       mEndStream = endStream;
       @State int originalState;
       @State int updatedState;
-      do {
-        originalState = mState.get();
-        updatedState = determineNextState(endStream, originalState, State.AWAITING_READ);
-      } while (!mState.compareAndSet(originalState, updatedState));
+      if (endStream && !data.hasRemaining()) {
+        // This read is the final read.
+        do {
+          originalState = mState.get();
+          updatedState = determineNextState(endStream, originalState, State.COMPLETE);
+        } while (!mState.compareAndSet(originalState, updatedState));
+        // onComplete still needs to be called - this always returns false.
+        mSucceededState.hasReachedFinalState(SucceededState.FINAL_READ_DONE);
+      } else {
+        do {
+          originalState = mState.get();
+          updatedState = determineNextState(endStream, originalState, State.AWAITING_READ);
+        } while (!mState.compareAndSet(originalState, updatedState));
```

Comment:
nit: I could be wrong, but it looks like the only difference between the two blocks of the if, is the call to `hasReachedFinalState()`. How 'bout we do this:

```
      bool isFinalRead = endStream && !data.hasRemaining();
      do {
        originalState = mState.get();
        updatedState = determineNextState(endStream, originalState, State.COMPLETE);
      } while (!mState.compareAndSet(originalState, updatedState));
      if (isFinalRead) {
        mSucceededState.hasReachedFinalState(SucceededState.FINAL_READ_DONE);
      }
```
to reduce some code duplication?